inst/include/mode/r/mode.hpp

```diff
@@ -57,18 +57,59 @@ cpp11::sexp mode_solve(SEXP ptr, double end_time) {
 double validate_time(cpp11::sexp r_time) {
   cpp11::doubles time = cpp11::as_cpp<cpp11::doubles>(r_time);
   if (time.size() != 1) {
-    cpp11::stop("expected 'time' to be a scalar value");
+    cpp11::stop("Expected 'time' to be a scalar value");
   }
   return time[0];
 }
 
+cpp11::integers object_dimensions(cpp11::sexp obj, size_t obj_size) {
+  cpp11::integers dim;
+  auto r_dim = obj.attr("dim");
+  if (r_dim == R_NilValue) {
+    dim = cpp11::writable::integers{static_cast<int>(obj_size)};
+  } else {
+    dim = cpp11::as_cpp<cpp11::integers>(r_dim);
+  }
+  return dim;
+}
+
+std::vector<double> validate_state(cpp11::sexp r_state,
+                              int n_state,
+                              int n_particles) {
+  cpp11::doubles r_state_data = cpp11::as_cpp<cpp11::doubles>(r_state);
+  const size_t state_len = r_state_data.size();
+  auto dim = object_dimensions(r_state, n_state);
+  if (dim.size() > 2) {
+    cpp11::stop("Expected 'state' to have at most 2 dimensions");
+  }
+  if (dim.size() == 2) {
+    if (dim[0] != n_state || dim[1] != n_particles) {
+      cpp11::stop("Expected 'state' to be a %d by %d matrix but was %d by %d",
+                  n_state, n_particles, dim[0], dim[1]);
+    }
+  }
+  if (dim.size() == 1 && state_len != n_state) {
+    cpp11::stop("Expected 'state' to be a vector of length %d but was length %d",
+                n_state, state_len);
+  }
+  std::vector<double> ret(state_len);
+  std::copy_n(REAL(r_state_data.data()), state_len, ret.begin());
+  return ret;
+}
+
 template <typename T>
-void mode_update_state(SEXP ptr, SEXP r_time) {
+void mode_update_state(SEXP ptr, SEXP r_state, SEXP r_time) {
   T *obj = cpp11::as_cpp<cpp11::external_pointer<T>>(ptr).get();
   if (r_time != R_NilValue) {
     auto time = validate_time(r_time);
     obj->set_time(time);
   }
+  if (r_state != R_NilValue) {
+    auto state = validate_state(r_state,
```

Comment:
It's not obvious in the dust code, but what it tries to do is validate everything first, then set everything (where nothing is expected to throw) so that either all components of state get successfully set or none to, avoiding inconsistent state. You could do this here by doing each if the `if`'s twice, but the scope gets pretty nasty. Might be easiest to try and pick this issue up later though once all the state elements are in